src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -78,27 +104,69 @@ etaConversionPass ast = do
 -- Function declarations                                                     --
 -------------------------------------------------------------------------------
 
--- | Applies 'etaConvertExpr' to the right-hand side of the given function
---   declaration.
+-- | Applies 'etaConvertTopLevelRhs' to the right-hand side of the given function
+--   declaration to ensure all functions and constructors on the right-hand side 
+--   are fully applied. 
+--   The missing top-level arguments are also added to the left-hand side of the 
+--   declaration and the function's type and the environment are updated accordingly. 
 etaConvertFuncDecl :: IR.FuncDecl -> Converter IR.FuncDecl
 etaConvertFuncDecl funcDecl = do
-  rhs' <- etaConvertExpr (IR.funcDeclRhs funcDecl)
-  return funcDecl { IR.funcDeclRhs = rhs' }
+  let rhs = IR.funcDeclRhs funcDecl
+  xs          <- generateFullArgumentList rhs
+  returnType' <- inEnv
+    $ lookupReturnType IR.ValueScope (IR.funcDeclQName funcDecl)
```

Comment:
I did that at first, but then the unit tests didn't work because apparently, `IR.funcDeclReturnType` contains `Nothing` when you define a test function.

Reply:
Okay, that makes sense. However, in the test cases you could just annotate the return type and add to the preconditions of the pass that the function declarations must be type annotated.

```haskell
  "f :: Foo -> Foo = g Foo" `shouldEtaConvert` "f (y :: Foo) :: Foo = g Foo y"
```